PROJECT_GUIDELINES.md

```diff
@@ -0,0 +1,173 @@
+  tag: "calcite-example",
+  styleUrl: "calcite-example.scss",
+  shadow: true
+})
+export class CalciteExample {
+
+  // ...
+
+  guid: string = `calcite-example-${guid()}`;
+
+  render() {
+    const id = this.el.id || this.guid;
+    return (
+      <Host id={id}></Host>
+    );
+  }
+
+  // ...
+}
+```
+
+This will create a unique id attribute like `id="calcite-example-51af-0941-54ae-22c14d441beb"` which should have a VERY low collision change since guid() generates IDs with `window.crypto.getRandomValues`. If a user supplies an `id` this will respect the users `id`.
+
+##### i18n
+
+Components should require as a few text translations as possible. In general lets users supply text values via slots and attributes. The lets user handle translations with their apps.
+
+If you component involves formatting numbers or dates use the [`Intl` APIs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) for formating the display of numbers and dates in your component.
+
+To add RTL support to your components you should use the internal `getElementDir` helper to apply the `dir` attribute to your component. This means that your components `dir` attribute will always match the documents `dir`.
+
+```tsx
+import { Component, Host, Element, h} from "@stencil/core";
+import { getElementDir } from "../../utils/dom";
+
+@Component({
+  tag: "calcite-component",
+  styleUrl: "calcite-component.scss",
+  shadow: true
+})
+export class CalciteComponent {
+  @Element() el: HTMLElement;
+
+  // ...
+
+  render() {
+    const dir = getElementDir(this.el);
+
+    return (
+      <Host dir={dir}>
+        <!-- The rest of your component -->
+      </Host>
+    );
+  }
+}
+```
+
+You can then implement direction specific CSS with CSS variables:
+
+```scss
+:host {
+  --calcite-tabs-tab-margin-start: 1.25rem;
+  --calcite-tabs-tab-margin-end: 0;
+}
+
+:host([dir="rtl"]) {
+  --calcite-tabs-tab-margin-start: 0;
+  --calcite-tabs-tab-margin-end: 1.25rem;
+}
+```
+
+Your component and its child components can then use `var(--calcite-tabs-tab-margin-start)` to access their proper values based on the direction of the document.
+
+In future it will likely become necessary to provide sting translations for components. An example would be the `aria-label` for the `<calcite-modal>` close button. Initial research looks promising and we could likely implement one of these approaches and set a `lang` for each component similar to how we set `dir`.
+
+- https://medium.com/stencil-tricks/implementing-internationalisation-i18n-with-stencil-5e6559554117 and https://codesandbox.io/s/43pmx55vo9
+- https://github.com/ionic-team/ionic-stencil-conference-app/issues/69
+
+#### Testing
+
+Place test files in the component folder using the `component.e2e.ts` naming convention. Write testable code and pure functions, avoid and extract side effects. Run tests locally before making any pull requests, and please, document tests with instructions if needed.
```

Comment:
> Place test files in the component folder using the `component.e2e.ts` naming convention.

Instead of this, we can link directly to https://stenciljs.com/docs/end-to-end-testing.

---

"Components should have an automated test for any incoming feature/bug fix."

---

It'd be good to mention here that PRs will run tests and not allow merging unless they all pass.

---

> document tests with instructions if needed

I would drop this and focus more on encouraging expressive test cases and code to indicate intent. Comments should be used sparingly and when the previous can't be achieved cleanly.